Merge pull request #2 from josdejong/develop

paulobuchsbaum opened the pull request:
Fix of bug fixes in `rationalize.js`, also changing  `simplify.js` and `simplifyConstant.js` and more  2 bugs in `simplify.js` and `simplifyconstant.js` in order to be possible passing in `Travis` test. 

**Bugs in `simplifyConstant.js` and   `simplify.js`**

1. `simplifyConstant.js` - I've changed `new ConstantNode(stringNumber, 'number')` to `new ConstantNode(number)`

1.  `simplify.js` - Due to problems with a  number node with  string type,  I've added `!isNaN(node.value)))` in number type test condition 

**Bugs in `rationalize.js`**

3.   I've fixed  negative power exponents and decimals coefficients troubles. The decimals coefficients problem has led to the need to add a new feature in `simplify.js ` and `simplifyConstant.js` (next topic)

**New feature in `simplify.js` and `simplifyConstant.js`**

4.  New rule type (string), whose valid values are in `listCommStrings` new variable. The only string rule accepted so far is to turn off exact fraction conversion in `simplifyConstant.js`

paulobuchsbaum commented:
There are yet some errors that I'm debugging locally ...

paulobuchsbaum commented:
Now everything is OK.  The only complain is that I don't get run Mocha in my local environment.

josdejong commented:
Thanks a lot for the fixes and improvements Paulo, looks good!

I don't fully understand `isExactFract`. Is this meant as a public option for `simplify`? If so, shouldn't we pass it via an object with options like `simplify(expr, ..., {exactFractions: false})` instead of "disguised" as a rule?

> I've researched in Internet and they say that Node.js does not accept import statement. However, I've tried some months ago and there was no problem! It is a kind of mistery

Sorry for the inconvenience. One month ago we released v5.0.0 where we changed the code from ES5 to ES6+, and set up build scripts to compile modern JS (in `src`) back to safe, old JS (in `lib` and `dist`). Thanks to the babel transpilation to ES5 we can use the newest javascript features, including import/export. However this isn't yet supported by node.js, which means we have to transpile everything too before we can run it in node.js. This includes running the unit tests: we first have to transpile to ES5, then we can run it on node.js. To do this, you have to pass an extra argument `--require babel-core/register` to mocha (see package.json):

```
mocha test test-node --recursive --require babel-core/register
```

Or simply run `npm test` of course, that will also run the linter that we introduced in mathjs v5 too :)

paulobuchsbaum commented:
> I don't fully understand isExactFract. Is this meant as a public option for simplify? If so, shouldn't we pass it via an object with options like simplify(expr, ..., {exactFractions: false}) instead of "disguised" as a rule?

Yes, you are right, @josdejong. I'll make the changes accordingly.

paulobuchsbaum commented:
Done!  

Your suggestion is much better than mine,  but I had to change the parameter profile in `simplify.js`, that was long (12 different options ). 

Now there is a new parameter in `simplify` function that can be used to put additional options. For now, it accepts only a boolean property: `exactFractions`

josdejong commented on src/function/algebra/simplify.js:
```diff
@@ -57,7 +57,7 @@ function factory (type, config, load, typed, math) {
    * Syntax:
    *
-   *     simplify(expr)
+   *     simplify(expr)n
```
I think the added character `n` is a typo?

josdejong commented:
Thanks @paulobuchsbaum , looks good like this, extending `simplify` with an options argument works out really nicely!

I'm not sure if you know this already, but you can run the linter locally via `npm run lint`, and it is also executed when running `npm test`.

Here a few feedback points:

1. I was wondering about the new added rules, do they belong in `rationalize`, or should we move them to `simplify`? 

    ```
    {l: 'c1*n + n', r: '(c1+1)*n'}, // Joining constants
    {l: 'c1*n - c2*n', r: '(c1-c2)*n'}, // Joining constants
    {l: 'c1*n - n', r: '(c1-1)*n'}, // Joining constants
    {l: 'v/c', r: '(1/c)*v'}, // variable/constant (new!)
    {l: 'v/-c', r: '-(1/c)*v'}, // variable/constant (new!)
    ```

2. It's a minor thing, but I personally prefer variable names like `options` instead of short names like `opts`. I find it more descriptive and since we nowadays we have IDE's with autocompletion and plenty of RAM, a few bytes extra isn't a problem ;)

3. It may be more straigtforward and easier to extend later if we simply pass `options` to all rules, like:

    ```js
    res = rules[i](res, options)
    ```

    instead of:

    ```js
    if (rules[i] === simplifyConstant) {
      res = rules[i](res, exactFract)
    } else {
      res = rules[i](res)
    }
    ```

    what do you think?
 
4. Currently, the `simplifyConstant` handles the exactFraction option by moving it in a "global" constant (ie. defined outside of the function) when calling `simplifyConstant`, and indirectly using it in `_toNumber`. I think it will be more robust to pass it as an argument to `_toNumber`, so it's straigtforward to see which functions use/pass these options. And same as (3.) maybe we should simply pass `options` instead of `exactFractions` everywhere.

5. I think we should describe this new option in the docs of `simplify`, and probably create two simple extra examples demonstrating what `exactFractions` does. If you want I can help out here too and write the docs.

6. Thanks for working out unit tests for the new simplify options :+1: .

paulobuchsbaum commented:
1 - I was wondering about the new added rules, do they belong in rationalize, or should we move them to simplify?

_In my opinion the following rules from rulesFirst constant can become default rules in `simplify`:_
```

      {l: '(-n1)/(-n2)', r: 'n1/n2'}, 
      {l: '(-n1)*(-n2)', r: 'n1*n2'}, 
      {l: 'n1--n2', r: 'n1+n2'}, 
       {l: 'n1-n2', r: 'n1+(-n2)'}, 
      {l: 'c1*n + c2*n', r: '(c1+c2)*n'}, 
      {l: 'c1*n + n', r: '(c1+1)*n'}, 
      {l: 'c1*n - c2*n', r: '(c1-c2)*n'}
      {l: 'c1*n - n', r: '(c1-1)*n'}, 
      {l: 'v/c', r: '(1/c)*v'},
      {l: 'v/-c', r: '-(1/c)*v'},
      {l: '-v*-c', r: 'c*v'}, 
      {l: '-v*c', r: '-c*v'}, 
      {l: 'v*-c', r: '-c*v'}, 
      {l: 'v*c', r: 'c*v'}, 
      {l: '-(-n1*n2)', r: '(n1*n2)'},
      {l: '-(n1*n2)', r: '(-n1*n2)'},
      {l: '-(-n1+n2)', r: '(n1-n2)'}, 
      {l: '-(n1+n2)', r: '(-n1-n2)'}, 
      {l: '-(-n1/n2)', r: '(n1/n2)'}, 
      {l: '-(n1/n2)', r: '(-n1/n2)'}]
```
_What do you think about it?_

2- I personally prefer variable names like `options` instead of short names like `opts`

_Ok, I agree, done_

3- It may be more straigtforward and easier to extend later if we simply pass options to all rules ...

_Ok, done_

4- .... Currently, the simplifyConstant handles the `exactFraction` option by moving it in a "global" constant (ie. defined outside of the function) when calling `simplifyConstant` .... 

Answer:
I confess I'm not much of a purist. In this case, I find it somewhat complicated and artificial, because `simplifyconstant` calls `foldfraction` that calls `-tonumber` that calls '-exactFraction'  but also
`simplifyconstant` calls `foldfraction` that calls `-eval` that calls  `-tonumber` that calls  `-exactFraction` 
and `-tonumber` receives parameters with `apply` 

It handles 18 different lines! 

So  if I put `optionsGlobal ` as local variable in  `factory `function (`simplifyConstant` wrap) receiving `option `parameter (with a comment) and also put a comment in  `-exactFractions`, talling that we are using a non local variable, it hurts you so much? 

If not, there will be a long chain of calls (shown above) that need to be changed, only to pass `options` parameter through several levels

5 -I think we should describe this new option in the docs of simplify, and probably create two simple extra examples demonstrating what exactFractions does. ...

 _I agree. I can do it, and add some `rationalize` documentation near._

josdejong commented:
Thanks for the updates Paulo, I hope to check it out and reply on your feedback asap.

josdejong commented:
Sorry for the late reply @paulobuchsbaum. I wanted to take the time to reply on your comments, it is an interesting discussion, thanks.

1 - I think it would be nice to move as much rules as possible to `simplify`, so `simplify` is as powerful as possible and `rationalize` can focus on its own job. We can leave it for now though and do this in a separate PR, both fine with me.

4 - Thanks for explaining your view. I totally agree that in this case passing options gives a lot of overhead and feels like unnecessary overhead.

I think a difference between our views is that you see the file as a single function containing some internal sub-functions, whilst I (generally) would like every function to be standalone/pure as much as possible, so it can be reasoned about on its own and can be extracted in separate files any moment.

You clearly describe all the interactions between the (sub) functions. There _is_ a data flow between these functions. When creating `globalOptions` as a "global" variable, this data flow becomes obfuscated and hidden. For example looking at `_toNumber` would not give you any idea that it has to be called by a function setting `globalOptions` first before you can call `_extractFraction` which consumes this global variable. Basically, you need to keep the whole file `simplifyConstant.js` in mind if you want to work on one of the (sub) functions inside it. The functions don't stand alone. It results in code with a much higher complexity.

At this point, `simplifyConstant.js` is so small that you can still keep it all in your head, and this `globalOptions` may be acceptable. As soon as the file starts growing, or needs to be refactored, it will become painful. 

Trying to reduce code complexity is really important to me. Using pure functions and immutable data is of *huge* help there. In my experience, code which is easy to understand but too verbose can be irritating to refactor but it doesn't really hurt. On the other hand, these sort of globals and hidden data flows can grow into a complex spagetti code which no one dares to touch, this can hurt a _lot_.

Again, at this point the code is very small and still maneagable. But I think this can easily grow into a wrong direction, and I rather want to start setting the right course right from the start. I'm ok with doing this refactoring myself if you find it unnecessary - but I really rather want to spend 15 minutes refactoring now than hours of painful debugging some day in the future.

5 - Great!

paulobuchsbaum commented:
I try to change the code to make the `options` parameter  local, but additionally I had to change some calls to `foldFraction` using `map`  to `for` loops and also I try to change typed function  `_toNumber` for a new profile with 2 parameters. 

I had changed and added 31 lines (> 21 lines that I said above). I've found the middle of chain calls very long and weird. I imagine all code reviewers  asking. _ WTF options means?_  because it only make some sense in the leave of tree calls, when finally options is used.  In the end, many new errors have arisen and I've got so confused, that I've reverted all changes to previous state. I give in! 

However, I think that it is very unlikely that new options will be necessary in `simplifyConstant.js` to the point of turning the code into spaghetti.

Another day, I will transfer the rules from `rationalize.js` to `simplify.js` default rules and propose an addendum in the documentation.

josdejong commented:
Thanks for giving it a try refactoring the `options` argument @paulobuchsbaum . I will merge your fixes now and give the refactoring a try too. If it indeed gives more issues then it solves I guess we should keep it as is. I will also describe `options` in the docs of `simplify` and add an example.

> If all the auxiliary functions (_toNumber, _eval, foldFraction and _exactFraction) were internal to simplifyConstant function, its parameters could be used directly without assignment to a global variable. Even so, is this against your philosophy?

In mathjs we have that this situation for example with the dependency injection mechanism which passes `typed` and `config` and a some other things. In such a case I would say it's best to use the parameters that are already defined in the outer scope of the function and not pass them as argument to internal functions. I think the main reason for me would be that (a) these parameters are constants in the scope of this function, they will never ever change again unlike our current `options` "global", and (b) if you would pass them as arguments too, you would have two variables with the same name, shadowing each other, which is confusing. Does that sort of make sense?

josdejong commented:
I've refactored `globalOptions` into `options`, see https://github.com/josdejong/mathjs/commit/e296fdcfe6b08cdbc7b841c727eac4876d64f5cc. There is about 30 occurances of options now, which is a lot.

> I've realized that using options as a local variable passing down through tree of calls get hard to discover when the options parameter is effectively used and in any middle call (_eval(....)) it's a complete mistery the options meaning for any code reviewer.

I actually _do_ like the transparency that I see after the refactoring: before, you had no clue that `_eval` actually needed `options` because one of the functions it calls needs `options`. Now you clearly see it. 

I guess it's partly personal preference and I guess we simply have different opinions here, which is ok. Thanks for the constructive discussion and your flexibility.

paulobuchsbaum commented:
Great!  I've got the meaning of your refactoring.  I've commit an error in `apply` usage. Thanks for all debate.

harrysarson commented:
Nice work here :+1: These changes look great, I agree that parsing options around will make this code much easier for others to read and will make maintainance much smoother going forwards :smile:

josdejong commented:
@paulmasson your fixes and improvements are now available in v5.1.0. Thanks again for all your effort!

paulmasson commented:
You’d be welcome, but I think you mean @paulobuchsbaum

josdejong commented:
whoops :flushed:, sorry Paul. I mean @paulobuchsbaum instead

paulobuchsbaum commented:
:relaxed:, thank you, @josdejong, for the opportunity to interact and learn.
